src/nnbench/context.py

```diff
@@ -183,3 +187,256 @@ def __call__(self) -> dict[str, Any]:
         result["memory_unit"] = self.memunit
         # TODO: Lacks CPU cache info, which requires a solution other than psutil.
         return {self.key: result}
+
+
+class Context:
+    def __init__(self, context: dict[str, Any] | ContextProvider | "Context" | None = None) -> None:
+        self._ctx_dict: dict[str, Any] = {}
+        if context:
+            self.merge(context, inplace=True)
+
+    def keys(self) -> KeysView[str]:
+        return self._ctx_dict.keys()
+
+    def values(self) -> ValuesView[Any]:
+        return self._ctx_dict.values()
+
+    def items(self) -> ItemsView[str, Any]:
+        return self._ctx_dict.items()
+
+    def merge(
+        self,
+        other: "Context" | dict[str, Any] | ContextProvider,
+        this_ctx: dict[str, Any] | None = None,
+        inplace: bool = True,
+    ) -> "Context":
+        """
+        Merge another dictionary, Context, or ContextProvider into this Context.
+
+        Parameters
+        ----------
+        other : Context | dict[str, Any] | ContextProvider
+            The other context to merge into this one.
+        this_ctx : dict[str, Any] | None, optional
+            The target dictionary to merge into. Takes this classes context if None.
+        inplace : bool, optional
+            If True, modifies the current context in place. Otherwise, returns a new merged Context.
```

Comment:
We need to be super careful with this to avoid surprises.

Returning a different instance from an instance method seems odd. How does pandas do this? A different instance with a view on the original?

Reply:
Pandas returns None for actions performed inplace (`inplace=True`) and a copy of the object with the action performed on `inplace=False`. I would go with their design to remain consistent. Not returning `self` also prohibits reassigning on non inplace operations. Thus there is a visual distinction between the two ways when reading code.

Also they have `inplace=False` as a default consistently to prevent unintentional data modifications. People won't work in notebooks with nnbench so the risk of changeing state in a session is lower. Still, I would go with the pandas way to make inplace changes deliberate and to use a familiar API design pattern.